util/qc_executor.py

```diff
@@ -101,33 +102,36 @@ def qc_check(self, parameter, dataset):
             processid = os.fork()
             if processid == 0:
                 # child process
-                with os.fdopen(w, 'w') as w:
-                    try:
-                        os.close(r)
-                        # run the qc function
-                        results = getattr(module, function_name)(**local_qc_args.get(function_name))
-                        # convert the np.ndarray into a string for sending over pipe
-                        bytes = io.BytesIO()
-                        np.savez(bytes, x=results)
-                        results_string = bytes.getvalue()
-                        w.write(results_string)
-                    except (TypeError, ValueError) as e:
-                        log.exception('<%s> Failed to execute QC %s %r', self.request_id, function_name, e)
-                        w.write(EXCEPTION_MESSAGE)
-                    finally:
-                        # if w.flush() is not called (or w.close()), the parent will not get data for results_string 
-                        w.flush()
-                        # child process is done, don't let it stick around
-                        os._exit(0)
+                # don't use 'with os.fdopen' since we call os._exit(0) below
```

Comment:
I believe the problem you're attempting to solve is the `os._exit(0)` in the original code was inside the with statement, causing the with statement to not be allowed to perform the teardown process (thus leaving the file handle open). 

Since with is the standard way to manage files in Python, I think the most direct fix is to move `os._exit(0)` to the same indentation as the `with` statement. This will prevent leaving the file handle open and I believe it will also remove the need for `w.flush()`.

Reply:
Yes, I was trying to address the os._exit(0) preventing proper teardown (although I don't know what issues if any it caused -- the zombies were a result of the missing wait -- but I figured I'd fix it while I was modifying the code). I like your suggestion and will give it a try.